packages/js/product-editor/src/blocks/product-fields/images/edit.tsx

```diff
@@ -46,30 +64,134 @@ export function Edit( {
 
 	const blockProps = useWooBlockProps( attributes, {
 		className: classnames( {
-			'has-images': images.length > 0,
+			'has-images': Array.isArray( propertyValue )
+				? propertyValue.length > 0
+				: Boolean( propertyValue ),
 		} ),
 	} );
 
-	const toggleRemoveZone = () => {
-		setIsRemovingZoneVisible( ! isRemovingZoneVisible );
-	};
+	function orderImages( newOrder: JSX.Element[] ) {
+		if ( Array.isArray( propertyValue ) ) {
+			const memoIds = propertyValue.reduce< Record< string, Image > >(
+				( current, item ) => ( {
+					...current,
+					[ `${ item.id }` ]: item,
+				} ),
+				{}
+			);
+			const orderedImages = newOrder
+				.filter( ( image ) => image?.props?.id in memoIds )
+				.map( ( image ) => memoIds[ image?.props?.id ] );
 
-	const orderImages = ( newOrder: JSX.Element[] ) => {
-		const orderedImages = newOrder.map( ( image ) => {
-			return images.find(
-				( file ) => file.id === parseInt( image?.props?.id, 10 )
+			recordEvent(
+				'product_images_change_image_order_via_image_gallery'
 			);
-		} );
-		recordEvent( 'product_images_change_image_order_via_image_gallery' );
-		setImages( orderedImages as MediaItem[] );
-	};
+			setPropertyValue( orderedImages );
+		}
+	}
 
-	const onFileUpload = ( files: MediaItem[] ) => {
-		if ( files[ 0 ].id ) {
-			recordEvent( 'product_images_add_via_file_upload_area' );
-			setImages( [ ...images, ...files ] );
+	function uploadHandler( eventName: string ) {
+		return function handleFileUpload( upload: MediaItem | MediaItem[] ) {
+			recordEvent( eventName );
+
+			if ( Array.isArray( upload ) ) {
+				const images: Image[] = upload
+					.filter( ( image ) => image.id )
+					.map( ( image ) => ( {
+						id: image.id,
+						name: image.title,
+						src: image.url,
+						alt: image.alt,
+					} ) );
+				if ( upload[ 0 ]?.id ) {
+					setPropertyValue( [
+						...( propertyValue as Image[] ),
+						...images,
+					] );
+				}
+			} else if ( upload.id ) {
+				setPropertyValue( mapUploadImageToImage( upload ) );
```

Comment:
This doesn't work, it appears that `upload` is still an array within the variation template.